### ISSUE TITLE
The computation of fisher matrix

### ISSUE THREAD

**Issue by shixianwen** [kuc2477/pytorch-ewc#1]
The computation for fisher matrix should be the average of the gradient^2. But in your implementation. you did the average of the loss, then only use that averaged loss to calculate the gradient and estimate the fisher matrix. 
Is this a bug or the math is equivalent.
check the tensorflow version: https://github.com/ariseff/overcoming-catastrophic

**Comment by Klitter** [comment]
> The computation for fisher matrix should be the average of the gradient^2. But in your implementation. you did the average of the loss, then only use that averaged loss to calculate the gradient and estimate the fisher matrix.
> Is this a bug or the math is equivalent.
> check the tensorflow version: https://github.com/ariseff/overcoming-catastrophic
I think fisher matrix should be calculated by second derivative of likelihood, but I notices its just uses gradient^2 so that I am confused. Which is correct?

**Comment by Klitter** [comment]
> > The computation for fisher matrix should be the average of the gradient^2. But in your implementation. you did the average of the loss, then only use that averaged loss to calculate the gradient and estimate the fisher matrix.
> > Is this a bug or the math is equivalent.
> > check the tensorflow version: https://github.com/ariseff/overcoming-catastrophic
> > I think fisher matrix should be calculated by second derivative of likelihood, but I notices its just uses gradient^2 so that I am confused. Which is correct?

For computing easier, Martens (2010 ICML Deep learning via Hessian-free optimization) selected the
empirical Fisher information matrix F¯, which is a crude (and biased) approximation to the well-known Fisher information matrix (Amari, 1998)

**Comment by kuc2477** [comment]
@shixianwen You're right. Diagonals of the fisher matrix is an expected value of the squared gradients of log-likelihoods, not squared expected gradients of log-likelihoods. I've [fixed it](https://github.com/kuc2477/pytorch-ewc/commit/151d75cdf98f85741c02204d0c8c540ee9578c8c#diff-93ff136fae812392eb0f68d1ce89b7feR74) and migrated the project to PyTorch >= 1.0. Thanks for reporting.
